### ISSUE TITLE
Verification e-mail for each registration

### ISSUE THREAD

**Issue by matthijskooijman** [evolution-events/xerxes#37]
This helps to check that the e-mailaddress on file is still current, so we can actually reach people in the event preparations.

**Comment by matthijskooijman** [comment]
> Can they still change the mail-address that's attached to it?

Yup, but that address change must be verified again.

> But what if it's not? Do we cancel the registration? Does it stay "pending" until it has been verified?

I guess it will stay pending.

**Comment by matthijskooijman** [comment]
One implementation could be to store a "last verified timestamp" for an e-mail address, and require that that timestamp is recent enough. This prevents having to store a "verified" flag along with the registration. One caveat here is that the e-mail address might expire between preparing a registration and actually confirming it, but if we check the confirmation timestamp only before migrating the registration from "in-progress" to "draft completed" and not when migrating from "draft completed" to "registered", this should not be a problem.

**Comment by matthijskooijman** [comment]
Thinking more, an explicit "verified" flag is probably easier and more transparent, since that allows verifying the e-mail even before completing the registration preparation. Once this flag is set, it is never unset (even when changing your address, since that should be separately confirmed before actually changing the e-mail address).

**Comment by matthijskooijman** [comment]
The "reverify your e-mail" for a registration is probably best handled by ourselves, rather than trying to shoehorn Djang-allauth's verification to suit.

Django-allauth keeps multiple e-mail addresses per user, each of has a verified flag (no timestamp...) and one of which is primary. In practice, the primary e-mail will probably be used for communication, and the secondary ones only for login, forgot-your-password and perhaps when the primary one bounces. This means that it is important to verify the primary e-mail address during registration, not so much the others.

One challenge is how to handle e-mail address changes, especially during the registration process. Normally, django-allauth only allows verified e-mail addresses as primary, so the procedure for changing the primary is: 1) add new e-mail address, 2) verify it, 3) make it primary. This probably works normally, but inside the registration wizard, this might not be ideal. A user is offered his primary e-mail (or all of them?) to check, and if he changes the address and you want to handle verification in the background, what should you show in the meantime? If verification is not complete, the primary address is not changed, but there is currently no way to store the "intent to become primary once verified". Additionally, verifying an e-mail address does not automatically make it primary.

One way to handle this is to break out of the wizard to change the e-mail address, but that would make it less likely for users to actually fix their e-mail address (reverification does enforce the primary is still in use, but it might not be their preferred one). One way to fix this is to, when the primary e-mail is changed, interrupt the wizard and let it continue after the user clicks the verification link?

**Comment by matthijskooijman** [comment]
Or perhaps somehow enforce only one e-mail address (possibly a second one while verifying, that replaces the primary one once verified)? Just one address might be easier for users, OTOH forget-your-password on multiple addresses is also useful.

**Comment by matthijskooijman** [comment]
If we do not migrate user passwords (enforcing everyone to recover their password through e-mail, see #48), then this implicitly verifies all e-mail addresses already, so we might be able to postpone this until after MVP.

Note that for new user registrations, verification is also done as part of the registration, but I think we would still need to explicitly check whether the e-mail address is verified, which might need to be done before MVP.
